### PR TITLE
1726 - void datatype for 7.3.x

### PR DESCRIPTION
**Pull request by heyrutvik**
library user need to `import scalaz.Scalaz._` in order to use Void syntax.

### CODE REVIEW

File: core/src/main/scala/scalaz/Void.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package scalaz
+
+import Liskov.<~<
+
+trait VoidModule {
+  type Void
+
+  def absurd[A](v: Void): A
+
+  def unsafeVoid: Void
```
Comment (hrhino) [inline core/src/main/scala/scalaz/Void.scala]: What do we have `unsafeVoid` for, again?

Comment (heyrutvik) [reply core/src/main/scala/scalaz/Void.scala]: honestly, I don't know.. I copied it from @alexknvl's implementation. So, he might have better idea. :)

Comment (alexknvl) [reply core/src/main/scala/scalaz/Void.scala]: For example, to convince the compiler that `(Int === String) => Void`.

**Comment by heyrutvik** [comment]
@jdegoes Rather than creating independent `VoidModule` trait. You are asking for extending it from  `scalaz.zio.Void` in 7.3.x. So that way, 7.3.x will depend on `ZIO` and instance will be `ZIO`'s `Void` type. Correct me if I'm wrong.

**Comment by jdegoes** [comment]
Yes, this is exactly what I suggest. The same instances and helpers can be provided, only they will use a safe coercion that can be hidden behind the API.

**Comment by heyrutvik** [comment]
@alexknvl @jdegoes @ktonga need help..
I was thinking to make following changes in order to make scalaz 7.3 to depend on ZIO. 
https://github.com/heyrutvik/scalaz/blob/61032f1d8c0360e13f15508625b1445b18d55608/core/src/main/scala/scalaz/package.scala#L380
```
import scalaz.zio.{Void => ZVoid}
package object scalaz {
  val Void: VoidModule = VoidImpl
  type Void = ZVoid
}
```
https://github.com/heyrutvik/scalaz/blob/61032f1d8c0360e13f15508625b1445b18d55608/core/src/main/scala/scalaz/Void.scala#L5
```
trait VoidModule extends Void {
...
}
```

Is this what you have suggested?
If yes then there is a problem because `zio.Void` is `final` in its declaration. https://github.com/heyrutvik/scalaz-zio/blob/master/core/shared/src/main/scala/scalaz/zio/Void.scala#L9

thoughts?

**Comment by jdegoes** [comment]
@heyrutvik 

Mostly correct, but you would not extend `Void`. Instead you can have:

```scala
trait VoidModule {
   type Void
   def absurd[A](v: Void): A
}
trait VoidImpl extends VoidModule {
  type Void = ZVoid
  def absurd[A](v: Void): A = v.absurd[A]
}
...
val Void: VoidModule = new VoidImpl{}
type Void = Void.Void
```

**Comment by heyrutvik** [comment]
@jdegoes Travis is happy only with scala 2.12 and _you know why_ :)

File: core/src/main/scala/scalaz/Void.scala [hunk 2]
```diff
@@ -1,29 +1,15 @@
 package scalaz
 
-import Liskov.<~<
+import scalaz.zio.{Void => ZVoid}
```
Comment (ktonga) [inline core/src/main/scala/scalaz/Void.scala]: `ZioVoid` maybe? we are in scalaz, just the Z ain't enough :slightly_smiling_face:

**Comment by jdegoes** [comment]
@heyrutvik Hopefully ZIO soon supports 2.11, right?

File: core/src/main/scala/scalaz/Void.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package scalaz
+
+import scalaz.zio.{Void => ZioVoid}
+
+trait VoidModule {
+  type Void
+
```
Comment (jdegoes) [inline core/src/main/scala/scalaz/Void.scala]: I think maybe want we want to do here is expose the type.

So maybe: `type Void = ZioVoid`. Then delete `VoidImpl`.

The reason is that then Scalaz will have the "same" Void as ZIO, which will result in better interop.

Comment (heyrutvik) [reply core/src/main/scala/scalaz/Void.scala]: Okay then implementation will be provided by `VoidModule` itself. right?

so in package object we will have something like..
```
val Void: VoidModule = new VoidModule {}
type Void = Void.Void
```

Comment (jdegoes) [reply core/src/main/scala/scalaz/Void.scala]: Yes, I'd make it an object.

File: core/src/main/scala/scalaz/syntax/VoidSyntax.scala [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package scalaz.syntax
+
+import scalaz._
+
+trait VoidOps {
```
Comment (jdegoes) [inline core/src/main/scala/scalaz/syntax/VoidSyntax.scala]: This should no longer be needed since it's now a method on `Void`.

**Comment by xuwei-k** [comment]
> I would prefer that (1) Scalaz 7.3 depend on ZIO

As of now, Scalaz 7.3 support scala 2.11, 2.12, 2.13 milestone, scala-js and scala-native. That doesn't appeal to me much.

**Comment by jdegoes** [comment]
> As of now, Scalaz 7.3 support scala 2.11, 2.12, 2.13 milestone, scala-js and scala-native. That doesn't appeal to me much.

What if ZIO supports all these too?

The question is whether to have zio-interop depend on 7.3, or have 7.3 pull in zio.

**Comment by ktonga** [comment]
fyi scalaz/scalaz-zio#89

**Comment by jdegoes** [comment]
@heyrutvik We have discussed at length and it seems that `Nothing` is going to have to work. We will have to disable a compiler warning in order to take advantage of this.

**Pull request closed without merging**